FirebaseSessions/Sources/FirebaseSessions.swift

```diff
@@ -89,23 +98,103 @@ protocol SessionsProvider {
 
     super.init()
 
+    SessionsDependencies.dependencies.forEach { subscriberName in
+      self.subscriberPromises[subscriberName] = Promise<Void>.pending()
+    }
+
+    Logger.logDebug("Expecting subscriptions from: \(SessionsDependencies.dependencies)")
+
     self.initiator.beginListening {
-      // On each session start, first update Settings if expired
-      self.settings.updateSettings()
-      let session = self.sessionGenerator.generateNewSession()
-      // Generate a session start event only when session data collection is enabled and if the session is allowed to dispatch events
-      if self.settings.sessionsEnabled, session.shouldDispatchEvents {
-        let event = SessionStartEvent(sessionInfo: session, appInfo: self.appInfo)
-        DispatchQueue.global().async {
-          self.coordinator.attemptLoggingSessionStart(event: event) { result in
-          }
+      // Generating a Session ID early is important as Subscriber
+      // SDKs will need to read it immediately upon registration.
+      let sessionInfo = self.sessionGenerator.generateNewSession()
+
+      // Post a notification so subscriber SDKs can get an updated Session ID
+      self.notificationCenter.post(name: Sessions.SessionIDChangedNotificationName,
+                                   object: nil)
+
+      let event = SessionStartEvent(sessionInfo: sessionInfo, appInfo: self.appInfo)
+
+      // Wait until all subscriber promises have been fulfilled before
+      // doing any data collection.
+      all(self.subscriberPromises.values).then(on: .global(qos: .background)) { _ in
+        guard self.isAnyDataCollectionEnabled else {
+          Logger
+            .logDebug(
+              "Data Collection is disabled for all subscribers. Skipping this Session Event"
+            )
+          return
+        }
+
+        Logger.logDebug("Data Collection is enabled for at least one Subscriber")
+
+        // Fetch settings if they have expired. This muse happen after the check for
+        // data collection, but before the check for sessionsEnabled from Settings
+        // so that re-fetching can happen.
+        self.settings.updateSettings()
```

Comment:
`updateSettings()` is async, not sure from the comment if this **must** be done "before the check for sessionsEnabled".

Reply:
Ah let me clarify - we need to make sure we fetch before we check if sessionsEnabled because if we didn't, we could get in a situation where settings disables itself indefinitely

Reply:
Got it, so `updateSettings()` doesn't necessarily need to complete before "check sessionsEnabled", it just has to be initiated?

Reply:
Yeah that's right

Reply:
I'm thinking if updateSettings should be done before waiting for all the dependencies initialization. What is the rationale to wait for all the dependencies to be initialized?

Reply:
> 

Yes, that is totally a valid point. Agreed! This makes sense now.